[JIT] Fix - Do not remove `CAST` nodes on store if the `LCL_VAR` is address exposed

TIHan opened the pull request:
## Description

Fixes:
- https://github.com/dotnet/runtime/issues/85081
- https://github.com/dotnet/runtime/issues/85382

On both ARM and XARCH, `mov` instructions get elided for `GT_STORE_LCL_VAR` when the destination and source registers are the same. This poses a problem when `GT_STORE_LCL_VAR` is for a NormalizedOnLoad **parameter**  and **struct field** variable and its `op` had its `CAST` node removed - which means sign/zero-extended instructions never get emitted when they are supposed to.

TIHan committed:
Do not remove CAST nodes on assignment if the LCL_VAR is a parameter.

SingleAccretion commented on src/coreclr/jit/morph.cpp:
```diff
@@ -10189,9 +10189,14 @@ GenTree* Compiler::fgOptimizeCastOnAssignment(GenTreeOp* asg)
+    {
+        LclVarDsc* varDsc = lvaGetDesc(effectiveOp1->AsLclVarCommon()->GetLclNum());
+
+        // It is not safe to remove the cast for non-NormalizeOnLoad variables or parameters.
```
This should mention the "why" - i. e. that this is a quirk/workaround for VN's handling of NOL and that we know this is in fact not a complete fix.

TIHan replied:
I don't that these issues are with VN's handling of NOLs; I could be wrong - it's very hard to determine.

jakobbotsch replied:
If this issue isn't with VN's handling, then what is it with? We should understand the problem before we work around it.

SingleAccretion replied:
Well - we don't know about other bugs related to this, but if there are, surely we would want to find out what they are as well.

FWIW, example from #85382 looks very much like the VN issue.

TIHan replied:
> We should understand the problem before we work around it.

I tried to describe the issue in the description regarding the `mov` elision.

SingleAccretion replied:
> which means sign/zero-extended instructions never get emitted

This is not a problem by itself - the upper bits are presumed to be undefined. In other words, `STORE_LCL_VAR`s do not have an implicit truncation semantic.

TIHan replied:
> STORE_LCL_VARs do not have an implicit truncation semantic.

So that's interesting. We definitely do this today.

Is there a place where NormalizeOnLoad rules are written down? Perhaps I can include these rules as part of this PR.

SingleAccretion replied:
The rules are conceptually simple (though not written down anywhere AFAIK):

1) All small locals are actually `TYP_INT` locals.
2) NOL locals are such that not all definitions can be controlled by the compiler and so the upper bits can be undefined. For parameters this is the case because of ABI. For struct fields - because of padding. For address-exposed locals - because not all stores are direct.
3) Hence, all NOL uses (unless proven otherwise) are assumed in morph to have undefined upper bits and explicit casts have be inserted to "normalize" them back to conform to IL semantics.

This naturally leads to the conclusion that direct stores need not be normalizing - after all, the uses already must assume undefined bits.

jakobbotsch replied:
I think the problem arises when morph concludes that it can forego normalizing in the following code:

https://github.com/dotnet/runtime/blob/f510c5ab2d8a9b94b84f63ff9d2d8318243e390f/src/coreclr/jit/morph.cpp#L4908-L4918

In this case the subrange assertion is coming from the previous
```scala
     ( 21, 24) [000106] -A--G---R--                         ▌  ASG       ushort
     (  4,  3) [000107] D------N---                         ├──▌  LCL_VAR   ushort V00 arg0         
     ( 16, 20) [000108] ----G------                         └──▌  CAST      int <- ushort <- long
     ( 15, 18) [000109] ----G------                            └──▌  RSH       long  
     (  3,  2) [000110] -----------                               ├──▌  LCL_VAR   long   V03 tmp1         
     (  8, 15) [000111] ----G------                               └──▌  AND       int   
     (  6, 13) [000112] n---G------                                  ├──▌  IND       byte  
     (  3, 10) [000113] H----------                                  │  └──▌  CNS_INT(h) long   0x7ff9d2f507ea static Fseq[s_17]
     (  1,  1) [000114] -----------                                  └──▌  CNS_INT   int    63
```
Then LSRA ends up allocating the cast and `V00` to the same register and the store that morph acted on is elided:
```scala
N085 (  9,  6) [000108] ----G------                  t108 = ▌  CAST      int <- long REG rax <l:$2cb, c:$2cc>
                                                            ┌──▌  t108   int    
N087 (  9,  6) [000107] DA--G--N---                         ▌  STORE_LCL_VAR ushort V00 arg0         d:3 rax REG rax $VN.Void
```

That morph code is quite questionable (and essentially puts NOL variables into very awkward territory, since now the frontend/middleend is suddenly reasoning about when NOLs are being allocated into registers). Not sure how bad it would be to remove.

jakobbotsch replied:
It seems likely that a peephole in the emitter would be able to get a large portion of the benefit of that subrange assertion check without having to complicate the NOL semantics with how stores are handled.

TIHan replied:
I did experiment with adding a peephole to look at redundant `movzx` instructions that would help certain cases.

@jakobbotsch so that opt, I disabled it and https://github.com/dotnet/runtime/issues/85382 did pass, but not https://github.com/dotnet/runtime/issues/85081.

Here is what I am using to test:
```
DOTNET_TieredCompilation=0
DOTNET_JitStressRegs=8
```
```csharp
using System.Runtime.CompilerServices;

public class Program
{
    public static long s_15;
    public static sbyte s_17;
    public static ushort s_21 = 36659;
    public static void Main()
    {
        s_15 = ~1;
        var x = (ushort)0;
        bool vr1 = M40(x);
    }

    public static bool M40(ushort arg0)
    {
        for (int var0 = 0; var0 < 2; var0++)
        {
            arg0 = 65535;
            arg0 &= (ushort)(s_15++ >> s_17);
            arg0 %= s_21;
        }

        System.Console.WriteLine(arg0);

        Test.MainTest();

        return false;
    }
}


public class Test
{
    public class Program
    {
        public static short s_2;

        [MethodImpl(MethodImplOptions.NoInlining)]
        public static void Consume(int x) { }

        [MethodImpl(MethodImplOptions.NoInlining)]
        public static int M8(byte arg0)
        {
            s_2 = 1;
            arg0 = (byte)(-s_2);
            var vr1 = arg0 & arg0;
            Consume(vr1);
            return vr1;
        }
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    public static int MainTest()
    {
        var x = (byte)1;
        var result = Test.Program.M8(x);
        if (result != 255)
        {
            return 0;
        }
        Console.WriteLine("Success");
        return 100;
    }
}
```

A successful run will print:
```
28876
Success
```

SingleAccretion replied:
So I too took a look at the dump for #85382 and the problem (in my interpretation) is that remorphing invalidates previous assumption as made by and acted on in assertion prop (@jakobbotsch - did you mean this too?):

1) We have `NOL = CAST<short>(...)`.
2) Local assertion prop takes a dependency on this, does not add cast for a downstream use.
3) Remorph invalidates the assumption by removing the cast (or, rather making it into a `CAST<int>` one).

jakobbotsch replied:
> (@jakobbotsch - did you mean this too?):

Yes, except I think the optimization made by local assertion prop makes the model of NOL locals much more complex than it should be (i.e. it means your explanation of NOL locals above is really not the fully story).
Local assertion prop here does not only rely on what it proves by seeing the subrange assertion. It also relies on normalization done by `genUnspillRegIfNeeded`, `GT_LCL_VAR` codegen and args homing to make this work. 
In other words: the assertion only meaningfully describes the upper bits if it just so happens that the local ended up being in a 4-byte home for the time between the assignment that created it and the use.

It's very subtle and hard to reason about this in the face of interacting optimizations like this one (another issue I dug up with an interacting optimization: #66624). So if we could get most of the benefits of this optimization through an emitter peephole, and simplify morph to always normalize, then I think that would be much preferable to the current state of things.

jakobbotsch replied:
FWIW, restricting the fix to parameters is not enough, it also needs to be restricted to struct fields. The following example fails in the same way:
```csharp
public static bool M40()
{
    S s = default;
    for (int var0 = 0; var0 < 2; var0++)
    {
        s.U = 65535;
        s.U &= (ushort)(s_15++ >> s_17);
        s.U %= s_21;
    }

    System.Console.WriteLine(s.U);
    return false;
}

public struct S { public ushort U; }
```

TIHan committed:
Added NormalizeOnLoad rules from SingleAccretion. Added description of why we cannot remove CAST nodes from parameters.

TIHan commented:
/azp run runtime-coreclr jitstress2-jitstressregs

TIHan committed:
Remove morph optimization for NormalizeOnLoad in fgMorphLocalVar. Update test.

TIHan commented:
@dotnet/jit-contrib this is ready. PTAL @jakobbotsch (even though you already have a little bit already :) )

I removed the NormalizeOnLoad optimization in `fgMorphLclVar` and I kept being pessimistic about parameters for removing casts on assignment. I also updated the test to include the use of struct fields.

On x64, regressions are:
+3,686 bytes, 534 regressions

When combined with https://github.com/dotnet/runtime/pull/85780, the regressions are (not including wins):
+1,729 bytes, 297 regressions

We don't have the same optimizations for ARM64 though.

TIHan commented:
/azp run runtime-coreclr jitstress2-jitstressregs

TIHan commented:
/azp run Fuzzlyn

TIHan commented:
@jakobbotsch - actually, let's wait on this. I think we need to have more discussions on what we are going to do here.

I saw this diff:
![image](https://user-images.githubusercontent.com/1278959/236573034-e916418e-10fb-4688-bdd8-d155d6e3d8e5.png)

And it makes me want to be careful how we go about this, on the other hand, it's not good that we have a correctness problem.

TIHan committed:
Merging

TIHan committed:
Merge remote-tracking branch 'upstream/main' into 85382-fix-2

TIHan committed:
Do not OptimizeCastOnStore for params and struct fields

TIHan commented:
/azp run runtime-coreclr jitstress2-jitstressregs

TIHan commented:
@dotnet/jit-contrib @jakobbotsch this is ready. Failures are unrelated. [Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=342759&view=ms.vss-build-web.run-extensions-tab) - There are regressions, which I don't think can be avoided safely since this is a correctness fix. However, I think we could expand peephole optimizations to cover some of these regressions in the future.

markples commented:
@tihan Was this an existing issue that was just found or fallout from recent CAST optimizations?  If the latter, then is the regression here (substantially?) less than the original win.  I assume so - just a check on the scope of the changes.  Thanks.

jakobbotsch commented on src/coreclr/jit/morph.cpp:
```diff
@@ -10115,8 +10115,14 @@ GenTree* Compiler::fgOptimizeCastOnStore(GenTree* store)
+    {
+        LclVarDsc* varDsc = lvaGetDesc(store->AsLclVarCommon()->GetLclNum());
+
+        // It is not safe to remove the cast for non-NormalizeOnLoad variables, parameters or struct fields.
+        if (!varDsc->lvNormalizeOnLoad() || varDsc->lvIsParam || varDsc->lvIsStructField)
```
This covers all normalize-on-load cases except OSR locals and address exposure cases. It should likely cover OSR locals as well. Additionally we should have a better comment, like @SingleAccretion has mentioned. Something like:

```suggestion
        // We can make this transformation only under the assumption that NOL locals are always normalized before they are used,
        // however this is not always the case: the JIT will utilize subrange assertions for NOL locals to make normalization
        // assumptions -- see fgMorphLeafLocal. Thus we can only do this for cases where we know for sure that subsequent uses
        // will normalize, which we can only guarantee when the local is address exposed.
        if (!varDsc->lvNormalizeOnLoad() || !varDsc->IsAddressExposed())
```

With this, I'd like to know what the benefit of this transformation even is. The address exposure case is likely very rare compared to the other cases that we are now disabling. How much do we actually gain from having this transformation with these restrictions?

jakobbotsch replied:
(I still question the logic in `fgMorphLeafLocal` -- I feel like it is nonsensical to rely on subrange assertions in the case that it does, which tightly couples it to implementation details of the backend. But that's probably something we can revisit in .NET 9, if we don't manage to find a better fix here.)

TIHan committed:
Update src/coreclr/jit/morph.cpp

Co-authored-by: Jakob Botsch Nielsen <Jakob.botsch.nielsen@gmail.com>

TIHan commented:
@markples It's not directly caused by any recent CAST optimizations, but those CAST optimizations may have allowed this particular issue around removing CAST around NOLs on STOREs(used to be ASG) to manifest itself.

TIHan committed:
Formatting

TIHan commented:
/azp run runtime-coreclr jitstress2-jitstressregs

TIHan commented:
@dotnet/jit-contrib @jakobbotsch This is ready again, less regressions than before and CI seems to be passing :)

jakobbotsch commented on src/coreclr/jit/morph.cpp:
```diff
@@ -10115,8 +10115,20 @@ GenTree* Compiler::fgOptimizeCastOnStore(GenTree* store)
+    {
+        LclVarDsc* varDsc = lvaGetDesc(store->AsLclVarCommon()->GetLclNum());
+
+        // We can make this transformation only under the assumption that NOL locals are always normalized before they
+        // are used,
+        // however this is not always the case: the JIT will utilize subrange assertions for NOL locals to make
+        // normalization
+        // assumptions -- see fgMorphLeafLocal. Thus we can only do this for cases where we know for sure that
+        // subsequent uses
+        // will normalize, which we can only guarantee when the local is address exposed.
```
The formatting here is a bit strange, maybe reformat the comment locally?

jakobbotsch commented on src/coreclr/jit/compiler.h:
```diff
@@ -1093,6 +1093,14 @@ class LclVarDsc
         return ((lvFldOffset % TARGET_POINTER_SIZE) == 0);
     }
+
+    // NormalizeOnLoad Rules:
+    //     1. All small locals are actually TYP_INT locals.
+    //     2. NOL locals are such that not all definitions can be controlled by the compiler and so the upper bits can
+    //        be undefined.For parameters this is the case because of ABI.For struct fields - because of padding.For
+    //        address - exposed locals - because not all stores are direct.
+    //     3. Hence, all NOL uses(unless proven otherwise) are assumed in morph to have undefined upper bits and
+    //        explicit casts have be inserted to "normalize" them back to conform to IL semantics.
```
Same here, this comment is formatted strangely, missing some spaces and it should be "address-exposed", not "address - exposed".

jakobbotsch reviewed (approved):
LGTM. Feel free to address the comment suggestions as part of a follow-up PR to avoid rerunning CI.
